File: imagepipeline-base/src/main/java/com/facebook/imageutils/WebpUtil.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
package com.facebook.imageutils;

import android.util.Pair;
import java.io.IOException;
import java.io.InputStream;
import javax.annotation.Nullable;

/**
 * This class contains utility method in order to manage the WebP format metadata
 */
public class WebpUtil {

  /**
   * Header for VP8 (lossy WebP). Take care of the space into the String
   */
  private static final String VP8_HEADER = "VP8 ";

  /**
   * Header for Lossless WebP images
   */
  private static final String VP8L_HEADER = "VP8L";

  /**
   * Header for WebP enhanced
   */
  private static final String VP8X_HEADER = "VP8X";

  private WebpUtil() {
  }

  /**
   * This method checks for the dimension of the WebP image from the given InputStream. We don't
   * support mark/reset and the Stream is always closed.
   *
   * @param is The InputStream used for read WebP data
   * @return The Size of the WebP image if any or null if the size is not available
   */
  @Nullable public static Pair<Integer, Integer> getSize(InputStream is) {
    // Here we have to parse the WebP data skipping all the information which are not
    // the size
    Pair<Integer, Integer> result = null;
    byte[] headerBuffer = new byte[4];
    try {
      is.read(headerBuffer);
      // These must be RIFF
      if (!compare(headerBuffer, "RIFF")) {
        return null;
      }
      // Next there's the file size
      getInt(is);
      // Next the WEBP header
      is.read(headerBuffer);
      if (!compare(headerBuffer, "WEBP")) {
        return null;
      }
      // Now we can have different headers
      is.read(headerBuffer);
      final String headerAsString = getHeader(headerBuffer);
      if (VP8_HEADER.equals(headerAsString)) {
        return getVP8Dimension(is);
      } else if (VP8L_HEADER.equals(headerAsString)) {
        return getVP8LDimension(is);
      } else if (VP8X_HEADER.equals(headerAsString)) {
        return getVP8XDimension(is);
      }
    } catch (IOException e) {
      e.printStackTrace();
    } finally {
      if (is != null) {
        try {
          is.close();
        } catch (IOException e) {
          e.printStackTrace();
        }
      }
    }
    // In this case we don't have the dimension
    return null;
  }

  /**
   * We manage the Simple WebP case
   *
   * @param is The InputStream we're reading
   * @return The dimensions if any
   * @throws IOException In case or error reading from the InputStream
   */
  private static @Nullable Pair<Integer, Integer> getVP8Dimension(final InputStream is)
      throws IOException {
    // We need to skip 7 bytes
    is.skip(7);
    // And then check the signature
    final short sign1 = getShort(is);
    final short sign2 = getShort(is);
    final short sign3 = getShort(is);
    if (sign1 != 0x9D || sign2 != 0x01 || sign3 != 0x2A) {
      // Signature error
      return null;
    }
    // We read the dimensions
    return new Pair<>(get2BytesAsInt(is), get2BytesAsInt(is));
  }

  /**
   * We manage the Lossless WebP case
   *
   * @param is The InputStream we're reading
   * @return The dimensions if any
   * @throws IOException In case or error reading from the InputStream
   */
  private static @Nullable Pair<Integer, Integer> getVP8LDimension(final InputStream is)
      throws IOException {
    // Skip 4 bytes
    getInt(is);
    //We have a check here
    final byte check = getByte(is);
    if (check != 0x2F) {
      return null;
    }
    int data1 = ((byte) is.read()) & 0xFF;
    int data2 = ((byte) is.read()) & 0xFF;
    int data3 = ((byte) is.read()) & 0xFF;
    int data4 = ((byte) is.read()) & 0xFF;
    // In this case the bits for size are 14!!! The sizes are -1!!!
    final int width = ((data2 & 0x3F) << 8 | data1) + 1;
    final int height = ((data4 & 0x0F) << 10 | data3 << 2 | (data2 & 0xC0) >> 6) + 1;
    return new Pair<>(width, height);
  }

  /**
   * We manage the Extended WebP case
   *
   * @param is The InputStream we're reading
   * @return The dimensions if any
   * @throws IOException In case or error reading from the InputStream
   */
  private static Pair<Integer, Integer> getVP8XDimension(final InputStream is) throws IOException {
    // We have to skip 8 bytes
    is.skip(8);
    // Read 3 bytes for width and height
    return new Pair<>(read3Bytes(is) + 1, read3Bytes(is) + 1);
  }

  /**
   * Compares some bytes with the text we're expecting
   *
   * @param what The bytes to compare
   * @param with The string those bytes should contains
   * @return True if they match and false otherwise
   */
  private static boolean compare(byte[] what, String with) {
    if (what.length != with.length()) {
      return false;
    }
    for (int i = 0; i < what.length; i++) {
      if (with.charAt(i) != what[i]) {
        return false;
      }
    }
    return true;
  }

  private static String getHeader(byte[] header) {
    StringBuilder str = new StringBuilder();
    for (int i = 0; i < header.length; i++) {
      str.append((char) header[i]);
    }
    return str.toString();
  }

  private static int getInt(InputStream is) throws IOException {
    byte byte1 = (byte) is.read();
    byte byte2 = (byte) is.read();
    byte byte3 = (byte) is.read();
    byte byte4 = (byte) is.read();
    return (byte4 << 24) & 0xFF000000 |
        (byte3 << 16) & 0xFF0000 |
        (byte2 << 8) & 0xFF00 |
        (byte1) & 0xFF;
  }

  public static int get2BytesAsInt(InputStream is) throws IOException {
    byte byte1 = (byte) is.read();
    byte byte2 = (byte) is.read();
    return (byte2 << 8 & 0xFF00) | (byte1 & 0xFF);
  }

  private static int read3Bytes(InputStream is) throws IOException {
    byte byte1 = getByte(is);
    byte byte2 = getByte(is);
    byte byte3 = getByte(is);
    return (((int) byte3) << 16 & 0xFF0000) |
        (((int) byte2) << 8 & 0xFF00) |
        (((int) byte1) & 0xFF);
  }

  private static short getShort(InputStream is) throws IOException {
    return (short) (is.read() & 0xFF);
  }

  private static byte getByte(InputStream is) throws IOException {
    return (byte) (is.read() & 0xFF);
  }

  private static boolean isBitOne(byte input, int bitIndex) {
    return ((input >> (bitIndex % 8)) & 1) == 1;
  }
}

File: imagepipeline-base/src/main/java/com/facebook/imageutils/WebpUtil.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.imageutils

import java.io.IOException
import java.io.InputStream

/** This class contains utility method in order to manage the WebP format metadata */
object WebpUtil {

  /** Header for VP8 (lossy WebP). Take care of the space into the String */
  private const val VP8_HEADER: String = "VP8 "
  /** Header for Lossless WebP images */
  private const val VP8L_HEADER: String = "VP8L"
  /** Header for WebP enhanced */
  private const val VP8X_HEADER: String = "VP8X"

  /**
   * This method checks for the dimension of the WebP image from the given InputStream. We don't
   * support mark/reset and the Stream is always closed.
   *
   * @param stream The InputStream used for read WebP data
   * @return The Size of the WebP image if any or null if the size is not available
   */
  @JvmStatic
  fun getSize(stream: InputStream): Pair<Int, Int>? {
    // Here we have to parse the WebP data skipping all the information which are not
    // the size
    val headerBuffer = ByteArray(4)
    try {
      stream.read(headerBuffer)
      // These must be RIFF
      if (!compare(headerBuffer, "RIFF")) {
        return null
      }
      // Next there's the file size
      getInt(stream)
      // Next the WEBP header
      stream.read(headerBuffer)
      if (!compare(headerBuffer, "WEBP")) {
        return null
      }
      // Now we can have different headers
      stream.read(headerBuffer)
      when (getHeader(headerBuffer)) {
        VP8_HEADER -> return getVP8Dimension(stream)
        VP8L_HEADER -> return getVP8LDimension(stream)
        VP8X_HEADER -> return getVP8XDimension(stream)
      }
    } catch (e: IOException) {
      e.printStackTrace()
    } finally {
      try {
        stream.close()
      } catch (e: IOException) {
        e.printStackTrace()
      }
    }
    // In this case we don't have the dimension
    return null
  }

  /**
   * We manage the Simple WebP case
   *
   * @param stream The InputStream we're reading
   * @return The dimensions if any
   * @throws IOException In case or error reading from the InputStream
   */
  @Throws(IOException::class)
  private fun getVP8Dimension(stream: InputStream): Pair<Int, Int>? {
    // We need to skip 7 bytes
    stream.skip(7)
    // And then check the signature
    val sign1 = stream.getNextByteAsInt()
    val sign2 = stream.getNextByteAsInt()
    val sign3 = stream.getNextByteAsInt()
    return if (sign1 != 0x9D || sign2 != 0x01 || sign3 != 0x2A) {
      // Signature error
      null
    } else {
      // We read the dimensions
      Pair(get2BytesAsInt(stream), get2BytesAsInt(stream))
    }
  }

  /**
   * We manage the Lossless WebP case
   *
   * @param stream The InputStream we're reading
   * @return The dimensions if any
   * @throws IOException In case or error reading from the InputStream
   */
  @Throws(IOException::class)
  private fun getVP8LDimension(stream: InputStream): Pair<Int, Int>? {
    // Skip 4 bytes
    getInt(stream)
    // We have a check here
    val check = stream.getNextByteAsInt()
    if (check != 0x2F) {
      return null
    }
    val data1 = stream.read() and 0xFF
    val data2 = stream.read() and 0xFF
    val data3 = stream.read() and 0xFF
    val data4 = stream.read() and 0xFF
    // In this case the bits for size are 14!!! The sizes are -1!!!
    val width = (data2 and 0x3F shl 8 or data1) + 1
    val height = (data4 and 0x0F shl 10 or (data3 shl 2) or (data2 and 0xC0 shr 6)) + 1
    return Pair(width, height)
  }

  /**
   * We manage the Extended WebP case
   *
   * @param stream The InputStream we're reading
   * @return The dimensions if any
   * @throws IOException In case or error reading from the InputStream
   */
  @Throws(IOException::class)
  private fun getVP8XDimension(stream: InputStream): Pair<Int, Int> {
    // We have to skip 8 bytes
    stream.skip(8)
    // Read 3 bytes for width and height
    return Pair(read3Bytes(stream) + 1, read3Bytes(stream) + 1)
  }

  /**
   * Compares some bytes with the text we're expecting
   *
   * @param what The bytes to compare
   * @param with The string those bytes should contains
   * @return True if they match and false otherwise
   */
  private fun compare(what: ByteArray, with: String): Boolean {
    if (what.size != with.length) {
      return false
    }
    return what.indices.none { i -> with[i].code.toByte() != what[i] }
  }

  private fun getHeader(header: ByteArray): String {
    val str = StringBuilder()
    for (i in header.indices) {
      str.append(Char(header[i].toUShort()))
    }
    return str.toString()
  }

  @Throws(IOException::class)
  private fun getInt(stream: InputStream): Int {
    val byte1 = stream.getNextByteAsInt()
    val byte2 = stream.getNextByteAsInt()
    val byte3 = stream.getNextByteAsInt()
    val byte4 = stream.getNextByteAsInt()
    return (byte4 shl 24) or (byte3 shl 16) or (byte2 shl 8) or byte1
  }

  @JvmStatic
  @Throws(IOException::class)
  fun get2BytesAsInt(stream: InputStream): Int {
    val byte1 = stream.getNextByteAsInt()
    val byte2 = stream.getNextByteAsInt()
    return (byte2 shl 8) or byte1
  }

  @Throws(IOException::class)
  private fun read3Bytes(stream: InputStream): Int {
    val byte1 = stream.getNextByteAsInt()
    val byte2 = stream.getNextByteAsInt()
    val byte3 = stream.getNextByteAsInt()
    return (byte3 shl 16) or (byte2 shl 8) or byte1
  }

  @Throws(IOException::class) private fun InputStream.getNextByteAsInt(): Int = read() and 0xFF
}
